src/twisted/internet/test/test_tcp.py

```diff
@@ -2054,6 +2054,101 @@ def connectionMade(self):
         # If test fails, reactor won't stop and we'll hit timeout:
         runProtocolsWithReactor(self, ListenerProtocol(), Client(), TCPCreator())
 
+    def test_doubleHalfCloseReactorCleanup(self):
+        """
+        If one side half-closes its connection, and then the other side of the
+        connection calls C{loseWriteConnection}, and then C{loseConnection} in
+        {writeConnectionLost}, the connection is closed correctly.
+
+        This rather obscure case used to fail.
+        See U{https://twistedmatrix.com/trac/ticket/3037}.
+        This also covers a code branch for ticket U{https://twistedmatrix.com/trac/ticket/9553}.
+
+        This is the generic tests for most reactors.
+        This tests is overwritten for IOCP reactor, as IOCP reactor doesn't
+        have getReaders() and getWriters().
+        """
+
+        @implementer(IHalfCloseableProtocol)
+        class ListenerProtocol(ConnectableProtocol):
+            def readConnectionLost(self):
+                # Force the existence of a writer in an attempt to cover #9553.
+                self.transport.startWriting()
+                self.transport.loseWriteConnection()
+
+            def writeConnectionLost(self):
+                self.transport.loseConnection()
+
+        class Client(ConnectableProtocol):
+            def connectionMade(self):
+                self.transport.loseConnection()
+
+        initial_readers = []
+
+        def reactorSetUp(reactor):
```

Comment:
runProtocolsWithReactor already provides a hook to customize the reactor:

```python
def reactorSetUp(self, reactor):
    pass

def buildReactor(self, *args, **kwargs):
    reactor = super().buildReactor()
    self.reactorSetUp(reactor)
    return reactor

test_doubleHalfCloseReactorCleanup(self):
    ...
    def reactorSetUp(reactor):
        ...

    self.reactorSetUp = reactorSetUp
```

Reply:
or a delegate proxy:

```python
@attr.s(frozen=True, slots=True, auto_attribs=True)
class SetupReactorBuilder:
    _reactorBuilder: ReactorBuilder
    _setupReactor: Callable[[IReactorBase, None]

    def buildReactor(self, *args, **kwargs):
        reactor = self._reactorBuilder.buildReactor(*args, **kwargs)
        self._setupReactor(reactor)
        return reactor

    def runReactor(self, *args, **kwargs):
        return self._reactorBuilder.runReactor(*args, **kwargs)

class TestFoo:
    def testFoo(self):
        reactor = runProtocolsWithReactor(
            SetupReactorBuilder(self, reactorSetUp), server, client, TCPCreator()
        )
```

Reply:
Thanks for pointing this out, @graingert !  I should have looked it up before giving my feedback.